File: jre.go
package main

import (
	"errors"
	tm "github.com/buger/goterm"
	"os"
	"path"
	"path/filepath"
	"runtime"
)

func GetJREDownloadURL() string {
	if runtime.GOOS == "windows" {
		return "https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.9.1%2B1/OpenJDK11U-jre_x86-32_windows_hotspot_11.0.9.1_1.zip"
	} else if runtime.GOOS == "linux" {
		return "https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.9.1%2B1/OpenJDK11U-jre_x64_linux_hotspot_11.0.9.1_1.tar.gz"
	} else {
		return "https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.9.1%2B1/OpenJDK11U-jre_x64_mac_hotspot_11.0.9.1_1.tar.gz"
	}
}

func GetBinaryName() string {
	if runtime.GOOS == "windows" {
		return "javaw.exe"
	} else if runtime.GOOS == "linux" {
		return "javaw"
	} else {
		return "javaw"
	}
}

func DownloadJRE(dest string) (string, error) {
	downloadURL := GetJREDownloadURL()
	return Download(downloadURL, dest)
}

func UnpackJRE(pkg string) error {
	extension := path.Ext(pkg)
	if extension == ".zip" {
		return Unzip(pkg, filepath.Dir(pkg))
	} else if extension == ".gz" {
		return Untar(pkg, filepath.Dir(pkg))
	} else {
		print("Unsupported extension " + extension)
	}
	return nil
}

func FindJava(powbotDirectory string) (string, error) {
	jreDirectory := filepath.FromSlash(powbotDirectory + "/jre/")
	if _, err := os.Stat(jreDirectory); err != nil {
		return "", err
	}

	var javaPath = ""

	err := filepath.Walk(jreDirectory, func(path string, info os.FileInfo, err error) error {
		if err == nil && info.Name() == GetBinaryName() {
			absolutePath, _ := filepath.Abs(path)
			javaPath = absolutePath
		}
		return nil
	})

	if err != nil {
		return "", err
	} else if len(javaPath) == 0 {
		return "", errors.New("java.exe not found in " + jreDirectory)
	}
	return javaPath, nil
}

func ObtainJRE(powbotDirectory string) (string, error) {
	jreDirectory := filepath.FromSlash(powbotDirectory + "/jre/")
	if err := CreateDirectory(jreDirectory); err != nil {
		tm.Println(tm.Color("\tCouldn't create directory for JRE at " + jreDirectory, tm.RED))
		tm.Flush()
		return "", err
	}

	jrePackage, err := DownloadJRE(jreDirectory)
	if err != nil {
		tm.Println(tm.Color("\tCouldn't download JRE to " + jreDirectory, tm.RED))
		tm.Flush()
		return "", err
	}

	err = UnpackJRE(jrePackage)
	if err != nil {
		tm.Println(tm.Color("\tCouldn't unpack JRE to " + jreDirectory, tm.RED))
		tm.Flush()
		return "", err
	}

	java, err := FindJava(powbotDirectory)
	if err != nil {
		tm.Println(tm.Color("\tCouldn't find the unpacked Java executable in " + jreDirectory, tm.RED))
		tm.Flush()
		return "", err
	}
	return java, nil
}

func EnsureJREPresent(powbotDirectory string) (string, error) {
	java, err := FindJava(powbotDirectory)
	if err != nil {
		tm.Println()
		tm.Println(tm.Color("\tBundled Java not found - downloading...", tm.YELLOW))
		tm.Flush()
		return ObtainJRE(powbotDirectory)
	}
	return java, nil
}

File: JRE.cs
using System;
using System.IO;
using System.Diagnostics;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace powbot_launcher_v2
{
    class JRE
    {

        public static string GetDirectory() {
            string homeFolder = HomeFolder.GetDirectory();
            return System.IO.Path.Combine(homeFolder, "jre");
        }

        public static string GetOrObtainJREBinary() {
            string dir = GetDirectory();
            if (!System.IO.Directory.Exists(dir)) {
                System.IO.Directory.CreateDirectory(dir);
            }
            string binary = FindBinary(dir);
            if (binary != null) {
                Console.WriteLine($"Java binary found at {binary}");
            } else {
                Console.WriteLine("Java binary not found, downloading...");
                binary = ObtainJRE();
            }

            return binary;
        }

        public static string ObtainJRE() {
            string downloadUrl = GetJREDownloadURL();
            string outputFile = System.IO.Path.Combine(GetDirectory(), downloadUrl.Substring(downloadUrl.LastIndexOf("/") + 1));

            using (var client = new System.Net.WebClient())
            {
                client.DownloadFile(downloadUrl, outputFile);
            }

            if (outputFile.EndsWith(".zip")) 
            {
                if (!Unzip(outputFile))
                {
                    throw new Exception($"{outputFile} - Failed to decompress file");
                }
            } else if (outputFile.EndsWith(".gz"))
            {
                if (!Untar(outputFile))
                {
                    throw new Exception($"{outputFile} - Failed to decompress file");
                }
            } else {
                throw new Exception($"{outputFile} - File type not supported");
            }

            return FindBinary(GetDirectory());
        }

        private static Boolean Unzip(string zipFile) {
            return Shell.Execute("unzip", GetDirectory(), false, new List<string> {zipFile.Substring(zipFile.LastIndexOf("/") + 1)});
        }
        private static Boolean Untar(string gzFile) {
            return Shell.Execute("tar", GetDirectory(), false, new List<string> {"-xvf", gzFile.Substring(gzFile.LastIndexOf("/") + 1)});
        }

        private static string GetBinaryName() {
            OperatingSystem os = Environment.OSVersion;
            PlatformID pid = os.Platform;
            switch (pid) 
            {
                case PlatformID.Win32NT:
                case PlatformID.Win32S:
                case PlatformID.Win32Windows:
                case PlatformID.WinCE:
                    return "javaw.exe";
                default:
                    return "java";
            }
        }

        private static string GetJREDownloadURL() {
            OperatingSystem os = Environment.OSVersion;
            PlatformID pid = os.Platform;
            switch (pid) 
            {
                case PlatformID.Win32NT:
                case PlatformID.Win32S:
                case PlatformID.Win32Windows:
                case PlatformID.WinCE:
		            return "https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.9.1%2B1/OpenJDK11U-jre_x86-32_windows_hotspot_11.0.9.1_1.zip";
                case PlatformID.MacOSX:
                case PlatformID.Unix:
		            return "https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.9.1%2B1/OpenJDK11U-jre_x64_mac_hotspot_11.0.9.1_1.tar.gz";
                default:
		            return "https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.9.1%2B1/OpenJDK11U-jre_x64_linux_hotspot_11.0.9.1_1.tar.gz";
            }
        }

        private static string FindBinary(string sDir) 
        {
            try 
            {

                foreach (string d in System.IO.Directory.GetDirectories(sDir)) 
                {
                    string binary = FindBinary(d);
                    if (binary != null) {
                        return binary;
                    }
                }
                foreach (string f in System.IO.Directory.GetFiles(sDir)) 
                {
                    if (f.EndsWith($"{GetBinaryName()}")) {
                        return f;
                    }
                }
            }
            catch (System.Exception e)
            {
                Console.WriteLine(e.ToString());
            }
            
            return null;
        }
    }
}
